### PR TITLE
feat!: modify delayLoad functionality

### PR DESCRIPTION
**Pull request by MoumitaM**
**Description**
- Delay load feature modified to delay not only the initialization of the SDK but also loading from the CDN
- Added support for load option

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: gatsby-ssr.js [hunk 1]
```diff
@@ -4,136 +4,60 @@ var _react = require("react");
...
+      controlPlaneUrl = pluginOptions.controlPlaneUrl,
+      delayLoad = pluginOptions.delayLoad,
+      delayLoadTime = pluginOptions.delayLoadTime,
+      manualLoad = pluginOptions.manualLoad,
+      loadType = pluginOptions.loadType,
+      useNewSDK = pluginOptions.useNewSDK,
+      sdkURL = pluginOptions.sdkURL,
+      useBetaSDK = pluginOptions.useBetaSDK,
+      _pluginOptions$loadOp = pluginOptions.loadOptions,
+      loadOptions = _pluginOptions$loadOp === undefined ? {} : _pluginOptions$loadOp;
```
Comment (saikumarrs) [inline gatsby-ssr.js:25]: This is unnecessary. We can club them together. It applies for another thing above as well.

File: gatsby-ssr.js [hunk 2]
```diff
@@ -4,136 +4,60 @@ var _react = require("react");
...
+
 
   var sdkSrc = "https://cdn.rudderlabs.com/v1/rudder-analytics.min.js";
-  if (sdkURL) sdkSrc = sdkURL;
-  else if (useBetaSDK) {
+  if (sdkURL) sdkSrc = sdkURL;else if (useBetaSDK) {
```
Comment (saikumarrs) [inline gatsby-ssr.js:29]: Looks like some formatting issue here.

File: gatsby-ssr.js [hunk 3]
```diff
@@ -4,136 +4,60 @@ var _react = require("react");
...
   var writeKey = process.env.NODE_ENV === "production" ? prodKey : devKey;
 
-  var loadConfig = "'" + writeKey + "', '" + dataPlaneUrl + "'";
-
-  if (loadOptions) {
-    // Override config URL if provided separately
-    loadOptions.configUrl = controlPlaneUrl || loadOptions.configUrl;
-    loadConfig += ", " + JSON.stringify(loadOptions);
+  if (controlPlaneUrl) {
+    loadOptions.configUrl = controlPlaneUrl;
```
Comment (saikumarrs) [inline gatsby-ssr.js:42]: We should never modify inputs.

File: src/gatsby-ssr.js [hunk 4]
```diff
@@ -40,28 +41,34 @@ exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
...
-  /*
-    if (controlPlaneUrl) {
-      return `'${writeKey}', '${dataPlaneUrl}', {configUrl: '${controlPlaneUrl}'}`;
-    } else {
-      return `'${writeKey}', '${dataPlaneUrl}'`;
-    }
-  }; */
+  if (controlPlaneUrl) {
+    loadOptions.configUrl = controlPlaneUrl;
+  }
```
Comment (saikumarrs) [inline src/gatsby-ssr.js:46]: We should never modify the inputs.

File: src/gatsby-ssr.js [hunk 5]
```diff
@@ -40,28 +41,34 @@ exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
...
+
+  const loadConfig = `'${writeKey}', '${dataPlaneUrl}',` + JSON.stringify(loadOptions);
```
Comment (saikumarrs) [inline src/gatsby-ssr.js:48]:
```suggestion
  const loadConfig = `'${writeKey}', '${dataPlaneUrl}', ${JSON.stringify(loadOptions)}`;
```

File: src/gatsby-ssr.js [hunk 6]
```diff
@@ -40,28 +41,34 @@ exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
...
   const snippet = `rudderanalytics=window.rudderanalytics=[];for(var methods=["load","page","track","identify","alias","group","ready","reset","getAnonymousId","setAnonymousId"],i=0;i<methods.length;i++){var method=methods[i];rudderanalytics[method]=function(a){return function(){rudderanalytics.push([a].concat(Array.prototype.slice.call(arguments)))}}(method)}
-  ${delayLoad || manualLoad ? `` : `rudderanalytics.load(${loadConfig})`};
+  var s = document.createElement("script");
+    s.type = "text/javascript";
+    s.src = "${sdkSrc}";
+    if ("${loadType}" == "async") {
+      s.async = true;
+    } else if ("${loadType}" == "defer") {
+      s.defer = true;
+    }
```
Comment (saikumarrs) [inline src/gatsby-ssr.js:58]: We don't need to add all this logic to the snippet string itself. Deduce the script string outside of this and append it.

File: src/gatsby-ssr.js [hunk 7]
```diff
@@ -40,28 +41,34 @@ exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
...
+const instantLoad = `${snippet}${delayLoad || manualLoad ? `` : `rudderanalytics.load(${loadConfig})`};`;
+
   const delayedLoader = `
       window.rudderSnippetLoaded = false;
       window.rudderSnippetLoading = false;
       window.rudderSnippetLoadedCallback = undefined;
       window.rudderSnippetLoader = function (callback) {
         if (!window.rudderSnippetLoaded && !window.rudderSnippetLoading) {
           window.rudderSnippetLoading = true;
+          ${snippet}
```
Comment (saikumarrs) [inline src/gatsby-ssr.js:71]: The snippet should be added inside the `loader` function.

File: src/gatsby-ssr.js [hunk 8]
```diff
@@ -40,28 +41,34 @@ exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
...
+    s.src = "${sdkSrc}";
+    if ("${loadType}" == "async") {
+      s.async = true;
+    } else if ("${loadType}" == "defer") {
+      s.defer = true;
+    }
+    document.head.appendChild(s);
 `;
 
+const instantLoad = `${snippet}${delayLoad || manualLoad ? `` : `rudderanalytics.load(${loadConfig})`};`;
```
Comment (saikumarrs) [inline src/gatsby-ssr.js:62]:
```suggestion
const instantLoader = `${snippet}${delayLoad || manualLoad ? `` : `rudderanalytics.load(${loadConfig})`};`;
```

**Comment by saikumarrs** [comment]
Note that `loadOptions` support was already present in the plugin.

File: src/gatsby-ssr.js (hunk 4 above)
Comment (MoumitaM) [reply src/gatsby-ssr.js:46]: That is how it is being done in gatsby. So, I didn't change that functionality.

Comment (saikumarrs) [reply src/gatsby-ssr.js:46]: That's incorrect.

Comment (MoumitaM) [reply src/gatsby-ssr.js]: Addressed

File: src/gatsby-ssr.js (hunk 6 above)
Comment (MoumitaM) [reply src/gatsby-ssr.js]: Addressed

File: src/gatsby-ssr.js (hunk 7 above)
Comment (MoumitaM) [reply src/gatsby-ssr.js]: Addressed

File: gatsby-ssr.js [hunk 9]
```diff
@@ -18,39 +18,39 @@ exports.onRenderBody = function (_ref, pluginOptions) {
...
 
 
-  var sdkSrc = "https://cdn.rudderlabs.com/v1/rudder-analytics.min.js";
-  if (sdkURL) sdkSrc = sdkURL;else if (useBetaSDK) {
-    sdkSrc = "https://cdn.rudderlabs.com/v1.1/beta/rudder-analytics.min.js";
-  } else if (useNewSDK) {
-    sdkSrc = "https://cdn.rudderlabs.com/v1.1/rudder-analytics.min.js";
-  }
+  var sdkSrc = "https://cdn.rudderlabs.com/v1.1/rudder-analytics.min.js";
+  if (sdkURL) sdkSrc = sdkURL;
```
Comment (saikumarrs) [inline gatsby-ssr.js:27]:
```suggestion
  var sdkSrc = sdkURL || "https://cdn.rudderlabs.com/v1.1/rudder-analytics.min.js";
```

File: gatsby-ssr.js [hunk 10]
```diff
@@ -18,39 +18,39 @@ exports.onRenderBody = function (_ref, pluginOptions) {
...
   if (!prodKey || prodKey.length < 10) console.error("Your RudderStack prodKey must be at least 10 char in length.");
 
   if (devKey && devKey.length < 10) console.error("If present, your RudderStack devKey must be at least 10 char in length.");
 
   var writeKey = process.env.NODE_ENV === "production" ? prodKey : devKey;
 
-  if (controlPlaneUrl) {
-    loadOptions.configUrl = controlPlaneUrl;
-  }
+  loadOptions.configUrl = controlPlaneUrl || loadOptions.configUrl;
```
Comment (saikumarrs) [inline gatsby-ssr.js:35]:
```suggestion
  const loadOpts = { ...loadOptions, configUrl: controlPlaneUrl || loadOptions.configUrl };
```

File: gatsby-ssr.js [hunk 11]
```diff
@@ -18,39 +18,39 @@ exports.onRenderBody = function (_ref, pluginOptions) {
...
 
-  var loadConfig = "'" + writeKey + "', '" + dataPlaneUrl + "'," + JSON.stringify(loadOptions);
+  var loadConfig = "'" + writeKey + "', '" + dataPlaneUrl + "', " + JSON.stringify(loadOptions);
```
Comment (saikumarrs) [inline gatsby-ssr.js:37]:
```suggestion
  var loadConfig = "'" + writeKey + "', '" + dataPlaneUrl + "', " + JSON.stringify(loadOpts);
```

**Review by saikumarrs — CHANGES_REQUESTED** [review]

**Comment by saikumarrs** [comment]
Bump the version to 2.0.0.

**Comment by saikumarrs** [comment]
Update the `README.md` to reflect the current state.

File: README.md [hunk 12]
```diff
@@ -58,32 +58,32 @@ plugins: [
 
       // If you need to proxy events through a custom data plane,
       // add a `dataPlaneUrl` property (defaults to https://hosted.rudderlabs.com )
-      // RudderStack docs:
-      //   - https://docs.rudderstack.com/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk#3-1-load
+      // Rudderstack docs:
```
Comment (saikumarrs) [inline README.md:61]: Replace all the instances of `Rudderstack` with `RudderStack`.

File: README.md [hunk 13]
```diff
@@ -58,32 +58,32 @@ plugins: [
...
 
       // Add a `controlPlaneUrl` property if you are self-hosting the Control Plane
-      // RudderStack docs:
-      //   - https://docs.rudderstack.com/rudderstack-sdk-integration-guides/rudderstack-javascript-sdk#3-1-1-self-hosted-control-plane
+      // Rudderstack docs:
+      //   - https://rudderstack.com/docs/sources/event-streams/sdks/rudderstack-javascript-sdk/#311-self-hosted-control-plane
       controlPlaneUrl: `https://override-control-plane-url`,
 
-      // boolean (defaults to false); whether to delay load RudderStack
+      // boolean (defaults to false); whether to delay loading(Download SDK and call load API) of RudderStack JS SDK.
```
Comment (saikumarrs) [inline README.md:70]:
```suggestion
      // boolean (defaults to false); whether to delay loading (add SDK script and load the SDK) of RudderStack JS SDK.
```

File: README.md [hunk 14]
```diff
@@ -95,6 +95,12 @@ plugins: [
       // *Another use case is if you want to add callbacks to the methods at load time.
       manualLoad: false,
 
+      // Can be used to override the location of the JavaScript library. This is useful
+      // if you want to add a proxy between the location and the RudderStack source so ad-blockers won't block  
+      // fetching the JavaScript sdk.
+      // By default plugin will use the latest JS SDK: https://cdn.rudderlabs.com/v1.1/rudder-analytics.min.js
+      sdkURL: `https://subdomain.yourdomain.com/v1.1/rudder-analytics.min.js`,
+
```
Comment (saikumarrs) [inline README.md:103]:
```suggestion
      // Can be used to override where the SDK should be loaded from. This is useful
      // if you want to serve the SDK from a custom domain other than RudderStack to tackle ad-blockers
      // By default, the plugin will use the latest JS SDK from RudderStack's CDN
      sdkURL: `https://subdomain.yourdomain.com/v1.1/rudder-analytics.min.js`,

```

File: README.md [hunk 15]
```diff
@@ -103,6 +109,8 @@ plugins: [
 
       // Options to the `load` API
       // Note: The above `controlPlaneUrl` overrides the `configUrl` field in this object
+      // Here you can find all the loadOptions available for JS SDK - 
+      // https://www.rudderstack.com/docs/sources/event-streams/sdks/rudderstack-javascript-sdk/load-js-sdk/#loading-options
```
Comment (saikumarrs) [inline README.md:113]:
```suggestion
      // You can find all the available load options here,
      // https://www.rudderstack.com/docs/sources/event-streams/sdks/rudderstack-javascript-sdk/load-js-sdk/#loading-options
```

File: README.md (hunk 12 above)
Comment (MoumitaM) [reply README.md:61]: Done

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**